include/h2o/tunnel.h

```diff
@@ -27,9 +27,53 @@ extern "C" {
 #endif
 
 typedef struct st_h2o_tunnel_t h2o_tunnel_t;
+typedef struct st_h2o_tunnel_endpoint_t h2o_tunnel_endpoint_t;
+typedef void (*h2o_tunnel_endpoint_open_cb)(h2o_tunnel_t *tunnel, h2o_tunnel_endpoint_t *end);
+typedef void (*h2o_tunnel_endpoint_send_cb)(h2o_tunnel_t *tunnel, h2o_tunnel_endpoint_t *end, h2o_iovec_t *bufs, size_t bufcnt,
+                                        int is_final);
+typedef void (*h2o_tunnel_endpoint_on_peer_send_complete_cb)(h2o_tunnel_t *tunnel, h2o_tunnel_endpoint_t *end, h2o_tunnel_endpoint_t *peer);
+typedef void (*h2o_tunnel_endpoint_close_cb)(h2o_tunnel_t *tunnel, h2o_tunnel_endpoint_t *end, const char *err);
 
-h2o_tunnel_t *h2o_tunnel_establish(h2o_context_t *ctx, h2o_socket_t *sock1, h2o_socket_t *sock2, h2o_timeout_t *timeout);
-void h2o_tunnel_break(h2o_tunnel_t *tunnel);
+typedef struct st_h2o_tunnel_endpoint_callbacks_t {
+    /**
+     * called when tunnel is initialized
+     */
+    h2o_tunnel_endpoint_open_cb open;
```

Comment:
We should not use function names defined in ISO C / POSIX, because invocations like `callbacks->open(` will end up in an error if `open(...)` is defined as a macro in the standard headers.

The case applies to `open`, `send`, `close` that is defined in this callback.

The convention we use to avoid the issue is to append `_` to the names that we define.